modules/turnos/controller/agenda.ts

```diff
@@ -835,6 +835,200 @@ export function actualizarEstadoAgendas(start, end) {
+                    $arrayElemAt: ['$resultado', 0]
+                },
+                _id: 0
+            }
+        },
+        {
+            $unwind: '$resultado'
+        },
+        {
+            $project: {
+                _id: '$resultado.prestaciones._id',
+                conceptId: '$resultado.prestaciones.conceptId',
+                fsn: '$resultado.prestaciones.fsn',
+                semanticTag: '$resultado.prestaciones.semanticTag',
+                term: '$resultado.prestaciones.term'
+            }
+        }
+
+    ];
+    let prestaciones = await agendaModel.aggregate(pipelinePrestaciones);
+    return prestaciones;
+}
+
+// agendas con el primer turno disponible agrupadas por profesional
+export async function turnosDisponibles(prestacion, organizacion) {
+    let pipelineAgendas = [];
+    pipelineAgendas = [
+
+        {
+            $match: {
+                'organizacion._id': {
+                    $eq: organizacion
+                },
+                'bloques.turnos.horaInicio': { $gte: new Date(moment().format('YYYY-MM-DD HH:mm')) },
+                estado: 'publicada',
+                'tipoPrestaciones.conceptId': prestacion
+            }
+        },
+        {
+            $match:
+            {
+                $or: [
+                    {
+                        'bloques.restantesProgramados': {
+                            $gt: 0
+                        }
+                    },
+                    {
+                        $and: [
+                            {
+                                'bloques.restantesDelDia': {
+                                    $gt: 0
+                                }
+                            }, {
+                                horaInicio: { $lte: new Date(moment().endOf('day').format('YYYY-MM-DD HH:mm')) }
+                            }]
+                    }]
+
+            }
+        },
+        {
+            $match: {
+                estado: 'publicada',
+                'tipoPrestaciones.conceptId': prestacion
+            }
+        },
+        {
+            $unwind: '$bloques'
+        },
+        {
+            $match: {
+                $expr: {
+                    $or: [
+                        { $gt: ['$bloques.restantesProgramados', 0] },
+                        { $gt: ['$bloques.restantesDelDia', 0] }
+                    ]
+                }
+            }
+        },
+        {
+            $unwind: '$bloques.tipoPrestaciones'
+        },
+        {
+            $unwind: '$bloques'
+        },
+        {
+            $unwind: '$bloques.turnos'
+        },
+        {
+            $match: {
+                'bloques.tipoPrestaciones.conceptId': prestacion
+            }
+        },
+        {
+            $match: {
+                'bloques.turnos.estado': 'disponible'
+            }
+        },
+        {
+            $sort: {
+                'bloques.turnos.horaInicio': 1
+            }
+        },
+        {
+            $group: {
```

Comment:
Para tener en cuenta: 

https://jira.mongodb.org/browse/DOCS-9755
https://dba.stackexchange.com/questions/197434/mongodb-is-order-guaranteed-deep-into-the-aggregation-pipeline

El $group no garantiza el orden de los elementos. Seguramente en una base de prueba con pocos datos puede que llegue bien ordenado. Pero en la base de producción, con muchas consultas de por medio puede que el resultado cambie improvisamente. 

Sugerencia: lo charlamos.

Reply:
dale!